flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java

```diff
@@ -163,6 +164,7 @@ public void sendSourceEventToCoordinator(SourceEvent event) {
 	@Override
 	public void close() throws Exception {
 		eventTimeLogic.stopPeriodicWatermarkEmits();
+		currentMainOutput.emitWatermark(Watermark.MAX_WATERMARK);
```

Comment:
Unfortunately that will not work in the `BATCH` mode we will introduce shortly. In `BATCH` mode the `currentMainOutput` will ignore all watermarks. At the same time we should still emit `MAX_WATERMARK` in the `BATCH` mode.

Reply:
Thanks for pointing out the problem and the detailed explaination. I have made the change at 406f376, please take a look. 
I also think the the SourceoperatorStreamTask should override the `advanceToEndOfEventTime`. If @aljoscha can confirm that, I'd love to made that change in this pr as well.